File: src/main/java/com/haysmike/MFTwisterExtension.java
package com.haysmike;

import com.bitwig.extension.controller.ControllerExtension;
import com.bitwig.extension.controller.api.*;

public class MFTwisterExtension extends ControllerExtension {
    private static final int maxMidiCcValue = 127;
    private static final int numControls = 64;

    private HardwareSurface hardwareSurface;

    protected MFTwisterExtension(final MFTwisterExtensionDefinition definition, final ControllerHost host) {
        super(definition, host);
    }

    @Override
    public void init() {
        final ControllerHost host = getHost();

        MidiIn midiIn = host.getMidiInPort(0);
        MidiOut midiOut = host.getMidiOutPort(0);

        hardwareSurface = host.createHardwareSurface();

        for (int i = 0; i < numControls; i++) {
            final int cc = i;
            RelativeHardwareKnob relativeHardwareKnob = hardwareSurface.createRelativeHardwareKnob("Knob " + cc);
            relativeHardwareKnob.setAdjustValueMatcher(midiIn.createRelativeBinOffsetCCValueMatcher(0, cc, maxMidiCcValue));
            // TODO: Look into `HardwareLight`
            relativeHardwareKnob.targetValue().addValueObserver(value -> midiOut.sendMidi(176, cc, (int) (value * maxMidiCcValue)));

            HardwareButton hardwareButton = hardwareSurface.createHardwareButton("Switch " + cc);
            HardwareAction pressedAction = hardwareButton.pressedAction();
            pressedAction.setActionMatcher(midiIn.createCCActionMatcher(1, cc, maxMidiCcValue));
            pressedAction.setBinding(host.createCallbackAction(() -> relativeHardwareKnob.setSensitivity(0.5), () -> "Handle switch pressed"));
            HardwareAction releasedAction = hardwareButton.releasedAction();
            releasedAction.setActionMatcher(midiIn.createCCActionMatcher(1, cc, 0));
            releasedAction.setBinding(host.createCallbackAction(() -> relativeHardwareKnob.setSensitivity(1.0), () -> "Handle switch released"));
        }
    }

    @Override
    public void exit() {
    }

    @Override
    public void flush() {
        hardwareSurface.updateHardware();
    }
}

File: src/main/java/com/haysmike/MFTwisterExtension.kt
package com.haysmike

import com.bitwig.extension.api.util.midi.ShortMidiMessage
import com.bitwig.extension.callback.IntegerValueChangedCallback
import com.bitwig.extension.callback.ShortMidiMessageReceivedCallback
import com.bitwig.extension.controller.ControllerExtension
import com.bitwig.extension.controller.api.ControllerHost
import kotlin.collections.HashSet
import kotlin.collections.MutableSet

class MFTwisterExtension(definition: MFTwisterExtensionDefinition?, host: ControllerHost?) :
    ControllerExtension(definition, host) {
    override fun init() {
        val host = getHost()

        val midiOut = host.getMidiOutPort(0)
        val userControlBank = host.createUserControls(numControls)
        for (i in 0..<numControls) {
            val cc = i
            val control = userControlBank.getControl(cc)
            val label = "CC " + cc
            control.setLabel(label)
            control.value().addValueObserver(midiRange, object : IntegerValueChangedCallback {
                override fun valueChanged(value: Int) {
                    midiOut.sendMidi(176, cc, value)
                }
            })
        }

        host.getMidiInPort(0).setMidiCallback(object : ShortMidiMessageReceivedCallback {
            override fun midiReceived(msg: ShortMidiMessage) {
                val cc = msg.getData1()
                val value = msg.getData2()
                if (msg.isControlChange() && msg.getChannel() == 1 && cc < numControls) {
                    if (value == 127) {
                        shiftedCcs.add(cc)
                    } else if (value == 0) {
                        shiftedCcs.remove(cc)
                    }
                } else if (msg.isControlChange() && msg.getChannel() == 0 && cc < numControls && isValidIncrementalValue(
                        value
                    )
                ) {
                    val resolution: Int =
                        if (shiftedCcs.contains(cc)) numControlStepsShifted else numControlStepsDefault
                    userControlBank.getControl(cc).inc(value - 64, resolution)
                }
            }
        })
    }

    override fun exit() {}

    override fun flush() {}

    private fun isValidIncrementalValue(value: Int): Boolean {
        return value == 63 || value == 65
    }

    companion object {
        private const val numControlStepsDefault = 128
        private const val numControlStepsShifted = 256
        private const val midiRange = 128
        private const val numControls = 64

        private val shiftedCcs: MutableSet<Int?> = HashSet<Int?>()
    }
}
